From 0185a6cbafee3210faf44140c34cb376d99f0691 Mon Sep 17 00:00:00 2001
From: Dave Stevenson <dave.stevenson@raspberrypi.org>
Date: Tue, 3 Apr 2018 16:06:46 +0100
Subject: [PATCH 57/80] staging: vc04_services: Split the VCHIQ MMAL layer out
 as a module

The VCHIQ MMAL layer is common with the bcm2835-v4l2 camera module.
Split it out as a module ready for the camera module to be adapted
to use the common library.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>
---
 drivers/staging/vc04_services/Kconfig         |  2 ++
 drivers/staging/vc04_services/Makefile        |  1 +
 .../vc04_services/bcm2835-codec/Kconfig       |  2 +-
 .../vc04_services/bcm2835-codec/Makefile      |  2 +-
 .../bcm2835-codec/bcm2835-v4l2-codec.c        | 13 +++++---
 .../staging/vc04_services/vchiq-mmal/Kconfig  |  7 ++++
 .../staging/vc04_services/vchiq-mmal/Makefile |  8 +++++
 .../mmal-common.h                             |  5 ---
 .../mmal-encodings.h                          |  0
 .../mmal-msg-common.h                         |  0
 .../mmal-msg-format.h                         |  0
 .../mmal-msg-port.h                           |  0
 .../{bcm2835-codec => vchiq-mmal}/mmal-msg.h  |  2 +-
 .../mmal-parameters.h                         |  0
 .../mmal-vchiq.c                              | 32 ++++++++++++++++---
 .../mmal-vchiq.h                              |  2 +-
 16 files changed, 58 insertions(+), 18 deletions(-)
 create mode 100644 drivers/staging/vc04_services/vchiq-mmal/Kconfig
 create mode 100644 drivers/staging/vc04_services/vchiq-mmal/Makefile
 rename drivers/staging/vc04_services/{bcm2835-codec => vchiq-mmal}/mmal-common.h (95%)
 rename drivers/staging/vc04_services/{bcm2835-codec => vchiq-mmal}/mmal-encodings.h (100%)
 rename drivers/staging/vc04_services/{bcm2835-codec => vchiq-mmal}/mmal-msg-common.h (100%)
 rename drivers/staging/vc04_services/{bcm2835-codec => vchiq-mmal}/mmal-msg-format.h (100%)
 rename drivers/staging/vc04_services/{bcm2835-codec => vchiq-mmal}/mmal-msg-port.h (100%)
 rename drivers/staging/vc04_services/{bcm2835-codec => vchiq-mmal}/mmal-msg.h (99%)
 rename drivers/staging/vc04_services/{bcm2835-codec => vchiq-mmal}/mmal-parameters.h (100%)
 rename drivers/staging/vc04_services/{bcm2835-codec => vchiq-mmal}/mmal-vchiq.c (98%)
 rename drivers/staging/vc04_services/{bcm2835-codec => vchiq-mmal}/mmal-vchiq.h (98%)

diff --git a/drivers/staging/vc04_services/Kconfig b/drivers/staging/vc04_services/Kconfig
index 06a6a163a317..6d592630ffc6 100644
--- a/drivers/staging/vc04_services/Kconfig
+++ b/drivers/staging/vc04_services/Kconfig
@@ -39,5 +39,7 @@ source "drivers/staging/vc04_services/bcm2835-codec/Kconfig"
 
 source "drivers/staging/vc04_services/vc-sm-cma/Kconfig"
 
+source "drivers/staging/vc04_services/vchiq-mmal/Kconfig"
+
 endif
 
diff --git a/drivers/staging/vc04_services/Makefile b/drivers/staging/vc04_services/Makefile
index 954d8553acfb..dfa39cf35899 100644
--- a/drivers/staging/vc04_services/Makefile
+++ b/drivers/staging/vc04_services/Makefile
@@ -15,6 +15,7 @@ obj-$(CONFIG_SND_BCM2835)	+= bcm2835-audio/
 obj-$(CONFIG_VIDEO_BCM2835)	+= bcm2835-camera/
 obj-$(CONFIG_BCM_VC_SM_CMA)	+= vc-sm-cma/
 obj-$(CONFIG_VIDEO_DECODE_BCM2835) += bcm2835-codec/
+obj-$(CONFIG_BCM2835_VCHIQ_MMAL) += vchiq-mmal/
 
 ccflags-y += -DVCOS_VERIFY_BKPTS=1 -Idrivers/staging/vc04_services -DUSE_VCHIQ_ARM -D__VCCOREVER__=0x04000000
 
diff --git a/drivers/staging/vc04_services/bcm2835-codec/Kconfig b/drivers/staging/vc04_services/bcm2835-codec/Kconfig
index 6bbb2a3ba10b..4656ca239c11 100644
--- a/drivers/staging/vc04_services/bcm2835-codec/Kconfig
+++ b/drivers/staging/vc04_services/bcm2835-codec/Kconfig
@@ -2,7 +2,7 @@ config VIDEO_DECODE_BCM2835
 	tristate "BCM2835 Video Decode support"
 	depends on MEDIA_SUPPORT
 	depends on VIDEO_V4L2 && (ARCH_BCM2835 || COMPILE_TEST)
-	select BCM2835_VCHIQ
+	select BCM2835_VCHIQ_MMAL
 	select VIDEOBUF2_VMALLOC
 	select V4L2_MEM2MEM_DEV
 	help
diff --git a/drivers/staging/vc04_services/bcm2835-codec/Makefile b/drivers/staging/vc04_services/bcm2835-codec/Makefile
index 4733cf2f9b5b..454f82ceb05d 100644
--- a/drivers/staging/vc04_services/bcm2835-codec/Makefile
+++ b/drivers/staging/vc04_services/bcm2835-codec/Makefile
@@ -1,5 +1,5 @@
 # SPDX-License-Identifier: GPL-2.0
-bcm2835-codec-objs := bcm2835-v4l2-codec.o mmal-vchiq.o
+bcm2835-codec-objs := bcm2835-v4l2-codec.o
 
 obj-$(CONFIG_VIDEO_DECODE_BCM2835) += bcm2835-codec.o
 
diff --git a/drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c b/drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c
index 5730bd300b01..5258e47218a2 100644
--- a/drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c
+++ b/drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c
@@ -35,10 +35,10 @@
 #include <media/v4l2-event.h>
 #include <media/videobuf2-dma-contig.h>
 
-#include "mmal-msg.h"
-#include "mmal-encodings.h"
-#include "mmal-parameters.h"
-#include "mmal-vchiq.h"
+#include "vchiq-mmal/mmal-msg.h"
+#include "vchiq-mmal/mmal-encodings.h"
+#include "vchiq-mmal/mmal-parameters.h"
+#include "vchiq-mmal/mmal-vchiq.h"
 
 MODULE_DESCRIPTION("BCM2835 codec V4L2 driver");
 MODULE_AUTHOR("Dave Stevenson, <dave.stevenson@raspberrypi.org>");
@@ -216,6 +216,11 @@ enum {
 	V4L2_M2M_DST = 1,
 };
 
+struct m2m_mmal_buffer {
+	struct v4l2_m2m_buffer	m2m;
+	struct mmal_buffer	mmal;
+};
+
 static inline struct bcm2835_codec_fmt_list *get_format_list(bool decode,
 							     bool capture)
 {
diff --git a/drivers/staging/vc04_services/vchiq-mmal/Kconfig b/drivers/staging/vc04_services/vchiq-mmal/Kconfig
new file mode 100644
index 000000000000..2288a95273e2
--- /dev/null
+++ b/drivers/staging/vc04_services/vchiq-mmal/Kconfig
@@ -0,0 +1,7 @@
+config BCM2835_VCHIQ_MMAL
+	tristate "BCM2835 MMAL VCHIQ service"
+	depends on (ARCH_BCM2835 || COMPILE_TEST)
+	select BCM2835_VCHIQ
+	help
+	  Enables the MMAL API over VCHIQ as used for the
+	  majority of the multimedia services on VideoCore.
diff --git a/drivers/staging/vc04_services/vchiq-mmal/Makefile b/drivers/staging/vc04_services/vchiq-mmal/Makefile
new file mode 100644
index 000000000000..f1647fc72afa
--- /dev/null
+++ b/drivers/staging/vc04_services/vchiq-mmal/Makefile
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: GPL-2.0
+bcm2835-mmal-vchiq-objs := mmal-vchiq.o
+
+obj-$(CONFIG_BCM2835_VCHIQ_MMAL) += bcm2835-mmal-vchiq.o
+
+ccflags-y += \
+	-Idrivers/staging/vc04_services \
+	-D__VCCOREVER__=0x04000000
diff --git a/drivers/staging/vc04_services/bcm2835-codec/mmal-common.h b/drivers/staging/vc04_services/vchiq-mmal/mmal-common.h
similarity index 95%
rename from drivers/staging/vc04_services/bcm2835-codec/mmal-common.h
rename to drivers/staging/vc04_services/vchiq-mmal/mmal-common.h
index 3d60796b4977..024d09583016 100644
--- a/drivers/staging/vc04_services/bcm2835-codec/mmal-common.h
+++ b/drivers/staging/vc04_services/vchiq-mmal/mmal-common.h
@@ -66,11 +66,6 @@ struct mmal_buffer {
 	s64 pts;
 };
 
-struct m2m_mmal_buffer {
-	struct v4l2_m2m_buffer	m2m;
-	struct mmal_buffer 	mmal;
-};
-
 /* */
 struct mmal_colourfx {
 	s32 enable;
diff --git a/drivers/staging/vc04_services/bcm2835-codec/mmal-encodings.h b/drivers/staging/vc04_services/vchiq-mmal/mmal-encodings.h
similarity index 100%
rename from drivers/staging/vc04_services/bcm2835-codec/mmal-encodings.h
rename to drivers/staging/vc04_services/vchiq-mmal/mmal-encodings.h
diff --git a/drivers/staging/vc04_services/bcm2835-codec/mmal-msg-common.h b/drivers/staging/vc04_services/vchiq-mmal/mmal-msg-common.h
similarity index 100%
rename from drivers/staging/vc04_services/bcm2835-codec/mmal-msg-common.h
rename to drivers/staging/vc04_services/vchiq-mmal/mmal-msg-common.h
diff --git a/drivers/staging/vc04_services/bcm2835-codec/mmal-msg-format.h b/drivers/staging/vc04_services/vchiq-mmal/mmal-msg-format.h
similarity index 100%
rename from drivers/staging/vc04_services/bcm2835-codec/mmal-msg-format.h
rename to drivers/staging/vc04_services/vchiq-mmal/mmal-msg-format.h
diff --git a/drivers/staging/vc04_services/bcm2835-codec/mmal-msg-port.h b/drivers/staging/vc04_services/vchiq-mmal/mmal-msg-port.h
similarity index 100%
rename from drivers/staging/vc04_services/bcm2835-codec/mmal-msg-port.h
rename to drivers/staging/vc04_services/vchiq-mmal/mmal-msg-port.h
diff --git a/drivers/staging/vc04_services/bcm2835-codec/mmal-msg.h b/drivers/staging/vc04_services/vchiq-mmal/mmal-msg.h
similarity index 99%
rename from drivers/staging/vc04_services/bcm2835-codec/mmal-msg.h
rename to drivers/staging/vc04_services/vchiq-mmal/mmal-msg.h
index 6a1c49ce3ca0..14ed1f0d0cb1 100644
--- a/drivers/staging/vc04_services/bcm2835-codec/mmal-msg.h
+++ b/drivers/staging/vc04_services/vchiq-mmal/mmal-msg.h
@@ -355,7 +355,7 @@ struct mmal_msg_port_parameter_get_reply {
 
 #define MMAL_EVENT_ERROR		MMAL_FOURCC('E', 'R', 'R', 'O')
 #define MMAL_EVENT_EOS			MMAL_FOURCC('E', 'E', 'O', 'S')
-#define MMAL_EVENT_FORMAT_CHANGED 	MMAL_FOURCC('E', 'F', 'C', 'H')
+#define MMAL_EVENT_FORMAT_CHANGED	MMAL_FOURCC('E', 'F', 'C', 'H')
 #define MMAL_EVENT_PARAMETER_CHANGED	MMAL_FOURCC('E', 'P', 'C', 'H')
 
 /* Structs for each of the event message payloads */
diff --git a/drivers/staging/vc04_services/bcm2835-codec/mmal-parameters.h b/drivers/staging/vc04_services/vchiq-mmal/mmal-parameters.h
similarity index 100%
rename from drivers/staging/vc04_services/bcm2835-codec/mmal-parameters.h
rename to drivers/staging/vc04_services/vchiq-mmal/mmal-parameters.h
diff --git a/drivers/staging/vc04_services/bcm2835-codec/mmal-vchiq.c b/drivers/staging/vc04_services/vchiq-mmal/mmal-vchiq.c
similarity index 98%
rename from drivers/staging/vc04_services/bcm2835-codec/mmal-vchiq.c
rename to drivers/staging/vc04_services/vchiq-mmal/mmal-vchiq.c
index 1d61313ad054..986b21e6a592 100644
--- a/drivers/staging/vc04_services/bcm2835-codec/mmal-vchiq.c
+++ b/drivers/staging/vc04_services/vchiq-mmal/mmal-vchiq.c
@@ -18,14 +18,15 @@
 
 #define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
 
+#include <linux/btree.h>
+#include <linux/completion.h>
 #include <linux/errno.h>
 #include <linux/kernel.h>
-#include <linux/mutex.h>
 #include <linux/mm.h>
+#include <linux/module.h>
+#include <linux/mutex.h>
 #include <linux/slab.h>
-#include <linux/completion.h>
 #include <linux/vmalloc.h>
-#include <linux/btree.h>
 #include <asm/cacheflush.h>
 
 #include "mmal-common.h"
@@ -40,6 +41,11 @@
 #define USE_VCHIQ_ARM
 #include "interface/vchi/vchi.h"
 
+MODULE_DESCRIPTION("BCM2835 MMAL VCHIQ interface");
+MODULE_AUTHOR("Dave Stevenson, <dave.stevenson@raspberrypi.org>");
+MODULE_LICENSE("GPL");
+MODULE_VERSION("0.0.1");
+
 /* maximum number of components supported */
 #define VCHIQ_MMAL_MAX_COMPONENTS 10
 
@@ -97,7 +103,7 @@ static const char *const port_action_type_names[] = {
 #if defined(FULL_MSG_DUMP)
 #define DBG_DUMP_MSG(MSG, MSG_LEN, TITLE)				\
 	do {								\
-		pr_debug(TITLE" type:%s(%d) length:%d\n",		\
+		pr_debug(TITLE " type:%s(%d) length:%d\n",		\
 			 msg_type_names[(MSG)->h.type],			\
 			 (MSG)->h.type, (MSG_LEN));			\
 		print_hex_dump(KERN_DEBUG, "<<h: ", DUMP_PREFIX_OFFSET,	\
@@ -111,7 +117,7 @@ static const char *const port_action_type_names[] = {
 #else
 #define DBG_DUMP_MSG(MSG, MSG_LEN, TITLE)				\
 	{								\
-		pr_debug(TITLE" type:%s(%d) length:%d\n",		\
+		pr_debug(TITLE " type:%s(%d) length:%d\n",		\
 			 msg_type_names[(MSG)->h.type],			\
 			 (MSG)->h.type, (MSG_LEN));			\
 	}
@@ -1657,6 +1663,7 @@ int vchiq_mmal_port_set_format(struct vchiq_mmal_instance *instance,
 
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_port_set_format);
 
 int vchiq_mmal_port_parameter_set(struct vchiq_mmal_instance *instance,
 				  struct vchiq_mmal_port *port,
@@ -1676,6 +1683,7 @@ int vchiq_mmal_port_parameter_set(struct vchiq_mmal_instance *instance,
 
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_port_parameter_set);
 
 int vchiq_mmal_port_parameter_get(struct vchiq_mmal_instance *instance,
 				  struct vchiq_mmal_port *port,
@@ -1692,6 +1700,7 @@ int vchiq_mmal_port_parameter_get(struct vchiq_mmal_instance *instance,
 
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_port_parameter_get);
 
 /* enable a port
  *
@@ -1722,6 +1731,7 @@ int vchiq_mmal_port_enable(struct vchiq_mmal_instance *instance,
 
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_port_enable);
 
 int vchiq_mmal_port_disable(struct vchiq_mmal_instance *instance,
 			    struct vchiq_mmal_port *port)
@@ -1742,6 +1752,7 @@ int vchiq_mmal_port_disable(struct vchiq_mmal_instance *instance,
 
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_port_disable);
 
 /* ports will be connected in a tunneled manner so data buffers
  * are not handled by client.
@@ -1829,6 +1840,7 @@ int vchiq_mmal_port_connect_tunnel(struct vchiq_mmal_instance *instance,
 
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_port_connect_tunnel);
 
 int vchiq_mmal_submit_buffer(struct vchiq_mmal_instance *instance,
 			     struct vchiq_mmal_port *port,
@@ -1875,6 +1887,7 @@ int vchiq_mmal_submit_buffer(struct vchiq_mmal_instance *instance,
 
 	return 0;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_submit_buffer);
 
 int mmal_vchi_buffer_init(struct vchiq_mmal_instance *instance,
 			  struct mmal_buffer *buf)
@@ -1888,6 +1901,7 @@ int mmal_vchi_buffer_init(struct vchiq_mmal_instance *instance,
 
 	return 0;
 }
+EXPORT_SYMBOL_GPL(mmal_vchi_buffer_init);
 
 int mmal_vchi_buffer_cleanup(struct mmal_buffer *buf)
 {
@@ -1910,6 +1924,7 @@ int mmal_vchi_buffer_cleanup(struct mmal_buffer *buf)
 #endif
 	return 0;
 }
+EXPORT_SYMBOL_GPL(mmal_vchi_buffer_cleanup);
 
 static void init_event_context(struct vchiq_mmal_instance *instance,
 			       struct vchiq_mmal_port *port)
@@ -2041,6 +2056,7 @@ int vchiq_mmal_component_init(struct vchiq_mmal_instance *instance,
 
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_component_init);
 
 /*
  * cause a mmal component to be destroyed
@@ -2071,6 +2087,7 @@ int vchiq_mmal_component_finalise(struct vchiq_mmal_instance *instance,
 
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_component_finalise);
 
 /*
  * cause a mmal component to be enabled
@@ -2096,6 +2113,7 @@ int vchiq_mmal_component_enable(struct vchiq_mmal_instance *instance,
 
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_component_enable);
 
 /*
  * cause a mmal component to be enabled
@@ -2121,6 +2139,7 @@ int vchiq_mmal_component_disable(struct vchiq_mmal_instance *instance,
 
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_component_disable);
 
 int vchiq_mmal_version(struct vchiq_mmal_instance *instance,
 		       u32 *major_out, u32 *minor_out)
@@ -2136,6 +2155,7 @@ int vchiq_mmal_version(struct vchiq_mmal_instance *instance,
 
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_version);
 
 int vchiq_mmal_finalise(struct vchiq_mmal_instance *instance)
 {
@@ -2163,6 +2183,7 @@ int vchiq_mmal_finalise(struct vchiq_mmal_instance *instance)
 
 	return status;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_finalise);
 
 int vchiq_mmal_init(struct vchiq_mmal_instance **out_instance)
 {
@@ -2250,3 +2271,4 @@ int vchiq_mmal_init(struct vchiq_mmal_instance **out_instance)
 	kfree(instance);
 	return -ENODEV;
 }
+EXPORT_SYMBOL_GPL(vchiq_mmal_init);
diff --git a/drivers/staging/vc04_services/bcm2835-codec/mmal-vchiq.h b/drivers/staging/vc04_services/vchiq-mmal/mmal-vchiq.h
similarity index 98%
rename from drivers/staging/vc04_services/bcm2835-codec/mmal-vchiq.h
rename to drivers/staging/vc04_services/vchiq-mmal/mmal-vchiq.h
index 372ed65f60c7..add7142b5e64 100644
--- a/drivers/staging/vc04_services/bcm2835-codec/mmal-vchiq.h
+++ b/drivers/staging/vc04_services/vchiq-mmal/mmal-vchiq.h
@@ -53,7 +53,7 @@ typedef void (*vchiq_mmal_buffer_cb)(
 struct vchiq_mmal_port {
 	bool enabled;
 	u32 handle;
-	u32 type; 	/* port type (enum vchiq_mmal_port_type), cached to use
+	u32 type;	/* port type (enum vchiq_mmal_port_type), cached to use
 			 * on port info set
 			 */
 	u32 index; /* port index, cached to use on port info set */
-- 
2.17.0

